File: cpp/969.pancake-sorting.cpp
/*
 * @lc app=leetcode id=969 lang=cpp
 *
 * [969] Pancake Sorting
 */
#include <vector>
#include <algorithm>
using namespace std;
// @lc code=start
// t:O(N^2), s:O(N)
class Solution
{
public:
    vector<int> pancakeSort(vector<int> &arr)
    {
        int w = arr.size();
        vector<int> res;
        int n = w;
        while (n)
        {
            if (arr.back() == n)
            {
                arr.pop_back();
            }
            else
            {
                vector<int> next_arr;
                int i = n - 1;
                for (; i != -1; --i)
                {
                    if (arr[i] == n)
                    {
                        if (i != 0)
                        {
                            res.push_back(i + 1);
                        }
                        res.push_back(n);
                        break;
                    }
                    next_arr.push_back(arr[i]);
                }
                for (int j = 0; j != i; ++j)
                {
                    next_arr.push_back(arr[j]);
                }
                arr = move(next_arr);
            }
            --n;
        }
        return res;
    }
};
// @lc code=end

File: 969.pancake-sorting.py
#
# @lc app=leetcode id=969 lang=python3
#
# [969] Pancake Sorting
#

# @lc code=start
class Solution:
    def pancakeSort(self, arr: List[int]) -> List[int]:
        w = len(arr)
        bucket_sort = [None]*w
        for i in range(w):
            bucket_sort[w-arr[i]] = i+1
        res = []
        for n in bucket_sort:
            for m in res:
                if n <= m:
                    n = m-n+1
            res.append(n)
            res.append(w)
            w -= 1
        return res

# @lc code=end
